ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs

```diff
@@ -15,17 +22,137 @@ module Ouroboros.Network.PeerSelection.LedgerPeers.Type
+
+instance ToJSON LedgerPeerSnapshot where
+  toJSON (LedgerPeerSnapshotV1 (slot, pools)) =
+    object [ "Version" .= (1 :: Int)
+           , "SlotNo" .= slot
+           , "BigLedgerPools" .= [ object [ "AccStake" .= show accStake
+                                          , "ReStake"  .= show stake
+                                          , "Relays"   .= relays]
+                                 | (AccPoolStake accStake, (PoolStake stake, relays)) <- pools]]
+
+instance FromJSON LedgerPeerSnapshot where
+  parseJSON = withObject "LedgerPeerSnapshot" $ \v -> do
+    vNum <- v .: "Version"
+    parsedSnapshot <-
+      case (vNum :: Int) of
+        1 -> do
+          slot <- v .: "SlotNo"
+          bigPools <- v .: "BigLedgerPools"
+          bigPools' <- (forM bigPools . withObject "BigLedgerPools" $ \poolV -> do
+            accStake <-
+              poolV .: "AccStake"
+              >>= (\case
+                    Just accStake -> pure accStake
+                    Nothing    ->
+                      fail "Invalid AccStake value")
+                  . readMaybe
+            reStake <-
+              poolV .: "ReStake"
+              >>= (\case
+                    Just reStake -> pure reStake
+                    Nothing    ->
+                      fail "Invalid ReStake value")
+                  . readMaybe
+            relays <- poolV .: "Relays"
+            return (accStake, (reStake, relays))) <?> Key "BigLedgerPools"
+
+          return $ LedgerPeerSnapshotV1 (slot, bigPools')
+        _ -> fail $ "Network.LedgerPeers.Type: parseJSON: failed to parse unsupported version " <> show vNum
+    case migrateLedgerPeerSnapshot parsedSnapshot of
+      Just payload -> return $ LedgerPeerSnapshot payload
+      Nothing      -> fail "Network.LedgerPeers.Type: parseJSON: failed to migrate big ledger peer snapshot"
+
+-- | cardano-slotting provides its own {To,From}CBOR instances for WithOrigin a
+-- but to pin down the encoding for CDDL we provide a wrapper with custom
+-- instances
+--
+newtype WithOriginWrapper = WithOriginWrapper (WithOrigin SlotNo)
+
+-- | Hand cranked CBOR instances to facilitate CDDL spec
+--
+instance ToCBOR WithOriginWrapper where
+  toCBOR (WithOriginWrapper Origin) = Codec.encodeListLen 1 <> Codec.encodeWord8 0
+  toCBOR (WithOriginWrapper (At (SlotNo slot))) = Codec.encodeListLen 2 <> Codec.encodeWord8 1 <> toCBOR slot
+
+instance FromCBOR WithOriginWrapper where
+  fromCBOR = do
+    listLen <- Codec.decodeListLen
+    tag <- Codec.decodeWord8
+    case (listLen, tag) of
+      (1, 0) -> pure $ WithOriginWrapper Origin
+      (1, _) -> fail "LedgerPeers.Type: Expected tag for Origin constructor"
+      (2, 1) -> WithOriginWrapper . At . SlotNo <$> fromCBOR
+      (2, _) -> fail "LedgerPeers.Type: Expected tag for At constructor"
+      _      -> fail "LedgerPeers.Type: Unrecognized list length while decoding WithOrigin SlotNo"
+
+instance ToCBOR LedgerPeerSnapshot where
```

Comment:
Do we have round trip tests for `{To,From}CBOR` instances?

Reply:
Yes.

Reply:
There's a test that checks if:
1. CBOR encoding is valid
2. CBOR round trip = identity
3. JSON decodes with the following stipulations:

   - since I've fully qualified every domain name, round trip equality test is modulo this
   - since serialisation of stake is in floating point, comparison to original Rational values is approximate